firmware/library/fx2conf.mk

```diff
@@ -31,6 +31,7 @@ ifeq ($(V),1)
 SDCCFLAGS += -V
 endif
 
-SDCC       = sdcc -mmcs51 $(SDCCFLAGS)
-SDAS       = sdas8051 -plo
-FX2LOAD    = python3 -m fx2.fx2tool -d $(VID):$(PID) load
+SDCC           = sdcc -mmcs51 $(SDCCFLAGS)
+SDCC_NEEDS_MF := $(shell expr `sdcc -v | sed -rn 's/.*([2-9]\.[0-9]+\.[0-9]+).*/\1/p'` \>= 4.2.9)
```

Comment:
That is cursed as hell. How is requiring this not a bug in sdcc?..

Reply:
I wish I knew. All I have to go on is [this mailing list thread](https://sourceforge.net/p/sdcc/mailman/sdcc-devel/thread/CAGjSPUDLCihL71_fv0Px4Ws_sa4E21Pq6yF0%2B6MSsecUFf-c1g%40mail.gmail.com/#msg37851015). And trying to build it, of course.

Phillip mentioned the reason:
> AFAIK, since GCC 3, 
this should be `-MMD -MF build/main.d` instead.
Current SDCC has rebased its GCC preprocessor fork on a newer upstream 
version, and apparently GCC has since dropped the pre-GCC-3 
compatibility syntax, and so did SDCC.

Contrary to what Phillip said in the rest of the message - 
> I haven't tried it, but I think 
`-MMD -MF build/main.d` should work both with current SDCC and any older 
SDCC from the last 15 years or so.

This does not work with SDCC 4.2.0, so I doubt it works with older versions.
Looking in [the manual](https://sdcc.sourceforge.net/doc/sdccman.pdf) in Section 1.5 Compatibility with previous versions, there is the following line:

> In 4.2.9, support for arguments to -MMD was dropped. The output file can be specified via -MF instead.

There is an alternative fix suggested by Felix -
> You can pass both -MMD and -MQ directly to sdcc of either version, there
> is no need for -Wp, if you can afford a second pass.
> 
> Try
> 
> $ sdcc -MQ .rel -MMD build/foo.c -o build/foo.d
> 
> Look at build/foo.d.

But I didn't manage to make it work without digging deep into the Makefile. Sorry!
I settled for the version check as the least intrusive fix (as I gave up on the other).

Further along in the mailing list, the reason for the rebasing and the need for the prototype fix is to work towards ISO C2X compliance.